internal/beater/beater.go

```diff
@@ -433,6 +462,25 @@ func (s *Runner) Run(ctx context.Context) error {
 	return result
 }
 
+var defaultDecoders = config.DefaultConfig().MaxConcurrentDecoders
+
+func maxConcurrentDecoders(memLimit float64, decoders uint) uint {
+	// If the decoders have already been set to a value different than the
+	// default configuration, return that value.
+	if decoders != defaultDecoders {
+		return decoders
+	}
+	if memLimit > 1 {
+		// Limit the number of concurrent decodes to 2048
+		max := uint(128 * memLimit)
+		if max > 2048 {
+			return 2048
+		}
+		return max
+	}
+	return defaultDecoders
+}
```

Comment:
A couple of things:
 - I think if a user sets a value explicitly, we should always use that. In the case where they explicitly set it to the default value, it shouldn't be changed based on memory.
 - I think we should assume memLimit(GB) is always valid (set to something conservative, in case of errors)

With those changes, we can get rid of `defaultDecoders` and simplify this:

```suggestion
func maxConcurrentDecoders(memLimitGB float64, decoders uint) uint {
	// If the decoders has been explicitly set, just return that value.
	if decoders > 0 {
		return decoders
	}
	// Allow 128 concurrent decoders for each 1GB memory, limited to at most 2048.
	const max = 2048
	decoders = uint(128 * memLimit)
	if decoders > max {
		decoders = max
	}
	return decoders
}
```